### PR TITLE
Added ProbeLevel to reduce memory overhead

### PR DESCRIPTION
**Pull request by pveentjer**
Added ProbeLevel to probes.

One can configure on each probe the probe level.

The Metricsregistry can be configured using a group property to have a minimum probe level (Mandatory by default). All probes that are registered with a too low probe level, are not being retained by the metrics registry. So you get some temporary litter, but the probe objects (especially strings) are not retained.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java [hunk 1]
```diff
@@ -100,6 +101,8 @@
      */
     PERFORMANCE_MONITOR_ENABLED("hazelcast.performance.monitoring.enabled", false),
 
+    PERFORMANCE_METRICS_LEVEL("hazelcast.performance.metric.level", ProbeLevel.MANDATORY.name()),
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java]: missing description of the property?

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/Probe.java [hunk 2]
```diff
@@ -68,4 +69,11 @@
      * @return the name of the Probe.
      */
     String name() default "";
+
+    /**
+     *
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/Probe.java]: Empty JavaDoc.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/SourceMetadata.java [hunk 3]
```diff
@@ -49,11 +49,15 @@
 
     void register(MetricsRegistryImpl metricsRegistry, Object source, String namePrefix) {
         for (FieldProbe field : fields) {
-            field.register(metricsRegistry, source, namePrefix);
+        //    if (field.probe.mandatory() || !metricsRegistry.trackMandatoryOnly) {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/SourceMetadata.java]: Leftover code?

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/ProbeLevelTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.hazelcast.internal.metrics;
+
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.internal.metrics.ProbeLevel.*;
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/ProbeLevelTest.java]: ProbeLevel.* import.

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/LongGaugeImplTest.java [hunk 5]
```diff
@@ -12,6 +13,8 @@
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import static com.hazelcast.internal.metrics.ProbeLevel.*;
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/LongGaugeImplTest.java]: ProbeLevel.* import.

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/ProbeLevelTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.internal.metrics.ProbeLevel.*;
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/ProbeLevelTest.java]: ProbeLevel.* import.

File: hazelcast/src/test/java/com/hazelcast/internal/monitors/HealthMonitorTest.java [hunk 7]
```diff
@@ -14,6 +15,7 @@
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import static com.hazelcast.internal.metrics.ProbeLevel.*;
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/monitors/HealthMonitorTest.java]: ProbeLevel.* import.

**Comment by Donnerbart** [comment]
I could not verify that all probes have the correct `ProbeLevel`, since I don't know all those internals. But this may change anyway over the time. The implementation itself looks fine.

:+1: (once green)

**Comment by jerrinot** [comment]
:+1:

**Comment by hasancelik** [comment]
:+1:

**Pull request merged**